File: scripts/producer/template_usage_approval.py
#!/usr/bin/env python3
"""Verify the controller-issued template-history approval before delivery."""
from __future__ import annotations

import json
import os
import re

from edit_scope import lane_required, resolve_scope, resolve_lanes
from fingerprints import file_sha256
from palmier.quality_hash import stable_hash

APPROVAL_FILE = ".sniper-template-usage-approved.json"
_SHA = re.compile(r"^[0-9a-f]{64}$")


def stored_operator_intent(producer_dir: str) -> dict:
    """Read immutable delivery authority from project.json, never plan.target."""
    candidates = [os.path.join(os.path.dirname(os.path.abspath(producer_dir)),
                               "project.json"),
                  os.path.join(os.path.abspath(producer_dir), "project.json")]
    project_path = next((item for item in candidates if os.path.isfile(item)), None)
    if not project_path:
        raise ValueError("delivery needs a valid stored operator intent in project.json")
    project = _json(project_path, "project intent")
    intent = project.get("resolvedIntent") or project.get("intent")
    if not isinstance(intent, dict) or intent.get("mode") not in ("short", "longform"):
        raise ValueError("delivery needs a valid stored operator intent in project.json")
    resolve_lanes(intent)
    return intent


def approval_required(producer_dir: str) -> bool:
    """Whether stored operator intent owns produced/full graphic choices."""
    intent = stored_operator_intent(producer_dir)
    return resolve_scope(intent) in ("produced", "full") \
        and lane_required(intent, "graphics")


def _json(path: str, label: str) -> dict:
    try:
        with open(path, encoding="utf-8") as handle:
            value = json.load(handle)
    except (OSError, json.JSONDecodeError) as exc:
        raise ValueError(f"cannot read {label} {path}: {exc}") from exc
    if not isinstance(value, dict):
        raise ValueError(f"{label} must be a JSON object")
    return value


def _sha(value: object, label: str) -> str:
    if not isinstance(value, str) or not _SHA.fullmatch(value):
        raise ValueError(f"{label} is not a SHA-256 digest")
    return value


def _transcript_digest(manifest: dict, manifest_path: str,
                       transcripts_dir: str) -> str:
    rows = []
    sources = manifest.get("sources") or []
    if not isinstance(sources, list):
        raise ValueError("manifest sources must be an array")
    for index, source in enumerate(sources):
        if not isinstance(source, dict):
            raise ValueError(f"manifest source {index} must be an object")
        source_id = str(source.get("id", index))
        transcript = source.get("transcriptPath")
        if transcript is None:
            rows.append([source_id, None, None])
            continue
        if not isinstance(transcript, str) or not transcript:
            raise ValueError(f"manifest source {source_id} has invalid transcriptPath")
        resolved = transcript if os.path.isabs(transcript) \
            else os.path.join(transcripts_dir, transcript)
        if not os.path.isfile(resolved):
            raise ValueError(f"referenced transcript is missing: {resolved}")
        rows.append([source_id, transcript, file_sha256(resolved)])
    return stable_hash(rows)


def _history_digest(producer_dir: str, history_path: str) -> str:
    root = os.path.realpath(os.path.join(producer_dir, ".sniper-learning", "runs"))
    target = os.path.realpath(history_path)
    try:
        inside = os.path.commonpath([root, target]) == root and target != root
    except ValueError:
        inside = False
    if not inside or os.path.islink(history_path) or not os.path.isfile(history_path):
        raise ValueError("template usage history is missing or unsafe")
    history = _json(history_path, "template usage history")
    found = _sha(history.get("digest"), "template usage history digest")
    core = {key: value for key, value in history.items() if key != "digest"}
    if found != stable_hash(core):
        raise ValueError("template usage history digest is invalid")
    return found


def _receipt(producer_dir: str) -> dict:
    path = os.path.join(producer_dir, APPROVAL_FILE)
    try:
        receipt = _json(path, "template usage approval")
    except ValueError as exc:
        raise ValueError("produced/full delivery needs a current template-history "
                         "review; review the saved plan first") from exc
    digest = _sha(receipt.get("digest"), "template usage approval digest")
    core = {key: value for key, value in receipt.items() if key != "digest"}
    if receipt.get("schemaVersion") != 1 \
            or receipt.get("kind") != "producer-template-usage-approval" \
            or digest != stable_hash(core):
        raise ValueError("template usage approval is malformed or tampered")
    return receipt


def require_current(plan_path: str, manifest_path: str, producer_dir: str,
                    transcripts_dir: str | None = None) -> None:
    """Fail closed unless exact plan/media/contract/history inputs were approved."""
    plan = _json(plan_path, "edit plan")
    intent = stored_operator_intent(producer_dir)
    if not approval_required(producer_dir):
        return
    manifest = _json(manifest_path, "asset manifest")
    receipt = _receipt(producer_dir)
    history_digest = _history_digest(producer_dir, str(receipt.get("historyPath")))
    contract = os.path.join(os.path.dirname(__file__), "template_usage_contract.py")
    operator_contract = os.path.join(os.path.dirname(__file__), "operator_intent_contract.py")
    current = {
        "planHash": file_sha256(plan_path),
        "planContentHash": stable_hash(plan),
        "manifestHash": file_sha256(manifest_path),
        "transcriptDigest": _transcript_digest(
            manifest, manifest_path, transcripts_dir or os.path.dirname(manifest_path)),
        "contractHash": file_sha256(contract),
        "operatorIntentDigest": stable_hash(
            {key: value for key, value in intent.items() if key != "preset"}),
        "operatorIntentContractHash": file_sha256(operator_contract),
    }
    stale = any(receipt.get(key) != value for key, value in current.items())
    if stale or receipt.get("historyDigest") != history_digest:
        raise ValueError("template-history approval is stale; review the current "
                         "saved plan before delivery")

File: src/lib/server/template-usage-approval.ts
import {
  existsSync, lstatSync, readFileSync, realpathSync, rmSync,
} from "node:fs";
import path from "node:path";
import { SCRIPTS_DIR } from "@/app/api/_lib/spawn-python";
import { readProjectJson } from "@/app/api/_lib/workspace";
import { resolveLanes, type ProjectIntent } from "@/lib/producer/intent-presets";
import { atomicWriteJsonSync } from "./atomic-file";
import { canonicalJsonSha256, fileSha256 } from "./auto-edit-hash";
import type { TemplateUsageAuthority } from "./template-usage-history";

export const TEMPLATE_USAGE_APPROVAL_FILE = ".sniper-template-usage-approved.json";
const TEMPLATE_CONTRACT = path.join(SCRIPTS_DIR, "producer", "template_usage_contract.py");
const OPERATOR_CONTRACT = path.join(SCRIPTS_DIR, "producer", "operator_intent_contract.py");
const SHA = /^[0-9a-f]{64}$/;

interface TemplateGateVerdict {
  ok: boolean;
  metrics?: Record<string, unknown>;
}

interface ApprovalInput {
  producerDir: string;
  planPath: string;
  manifestPath: string;
  transcriptsDir: string;
  templateUsage: TemplateUsageAuthority;
  verdict: TemplateGateVerdict;
  operatorIntentVerdict: TemplateGateVerdict;
}

interface DeliveryInput {
  producerDir: string;
  planPath: string;
  manifestPath: string;
  transcriptsDir?: string;
}

export interface TemplateUsageApproval {
  schemaVersion: 1;
  kind: "producer-template-usage-approval";
  planHash: string;
  planContentHash: string;
  manifestHash: string;
  transcriptDigest: string;
  contractHash: string;
  operatorIntentDigest: string;
  operatorIntentContractHash: string;
  historyPath: string;
  historyDigest: string;
  approvedAt: string;
  digest: string;
}

function hash(value: unknown): string {
  return canonicalJsonSha256(value);
}

function readJson(filePath: string): Record<string, unknown> {
  const value = JSON.parse(readFileSync(filePath, "utf8")) as unknown;
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${filePath} must contain a JSON object`);
  }
  return value as Record<string, unknown>;
}

function storedOperatorIntent(producerDir: string): ProjectIntent {
  const project = readProjectJson(path.dirname(path.resolve(producerDir)));
  const intent = project?.resolvedIntent ?? project?.intent;
  if (!intent || !["short", "longform"].includes(intent.mode)
      || !["trim", "light", "produced", "full"].includes(intent.scope)
      || !intent.lanes || typeof intent.lanes !== "object") {
    throw new Error("delivery needs a valid stored operator intent in project.json");
  }
  resolveLanes(intent.scope, intent.lanes);
  return intent;
}

function operatorIntentAuthority(producerDir: string): Omit<ProjectIntent, "preset"> {
  const authority = { ...storedOperatorIntent(producerDir) };
  delete authority.preset;
  return authority;
}

export function templateUsageApprovalRequired(producerDir: string): boolean {
  const intent = storedOperatorIntent(producerDir);
  return ["produced", "full"].includes(intent.scope)
    && resolveLanes(intent.scope, intent.lanes).graphics === "auto";
}

function transcriptDigest(manifestPath: string, transcriptsDir: string): string {
  const manifest = readJson(manifestPath);
  const sources = Array.isArray(manifest.sources) ? manifest.sources : [];
  const rows = sources.map((value, index) => {
    const source = value as Record<string, unknown>;
    const transcript = source.transcriptPath;
    if (transcript == null) return [String(source.id ?? index), null, null];
    if (typeof transcript !== "string" || !transcript) {
      throw new Error(`manifest source ${String(source.id ?? index)} has an invalid transcriptPath`);
    }
    const resolved = path.isAbsolute(transcript) ? transcript : path.join(transcriptsDir, transcript);
    const digest = fileSha256(resolved);
    if (!digest) throw new Error(`referenced transcript is missing: ${resolved}`);
    return [String(source.id ?? index), transcript, digest];
  });
  return hash(rows);
}

function historyDigest(producerDir: string, filePath: string): string {
  const root = path.resolve(producerDir, ".sniper-learning", "runs");
  const target = path.resolve(filePath);
  const relative = path.relative(root, target);
  if (!relative || relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error("template usage history path escapes its controller-owned run directory");
  }
  if (!existsSync(target) || lstatSync(target).isSymbolicLink() || !lstatSync(target).isFile()
      || path.relative(realpathSync(root), realpathSync(target)).startsWith("..")) {
    throw new Error("template usage history is missing or unsafe");
  }
  const history = readJson(target);
  const found = history.digest;
  const core = { ...history };
  delete core.digest;
  if (typeof found !== "string" || !SHA.test(found) || found !== hash(core)) {
    throw new Error("template usage history digest is invalid");
  }
  return found;
}

function content(input: DeliveryInput) {
  const plan = readJson(input.planPath);
  const transcriptsDir = input.transcriptsDir ?? path.dirname(input.manifestPath);
  const planHash = fileSha256(input.planPath);
  const manifestHash = fileSha256(input.manifestPath);
  const contractHash = fileSha256(TEMPLATE_CONTRACT);
  const operatorIntentContractHash = fileSha256(OPERATOR_CONTRACT);
  if (!planHash || !manifestHash || !contractHash || !operatorIntentContractHash) {
    throw new Error("template usage approval inputs or gate implementation are missing");
  }
  const operatorIntentDigest = hash(operatorIntentAuthority(input.producerDir));
  return {
    planHash,
    planContentHash: hash(plan),
    manifestHash,
    transcriptDigest: transcriptDigest(input.manifestPath, transcriptsDir),
    contractHash,
    operatorIntentDigest,
    operatorIntentContractHash,
  };
}

export function templateUsageApprovalPath(producerDir: string): string {
  return path.join(producerDir, TEMPLATE_USAGE_APPROVAL_FILE);
}

export function clearTemplateUsageApproval(producerDir: string): void {
  rmSync(templateUsageApprovalPath(producerDir), { force: true });
}

/** Commit the exact controller-bound history that passed the deterministic gate. */
export function writeTemplateUsageApproval(input: ApprovalInput): void {
  if (!templateUsageApprovalRequired(input.producerDir)) {
    clearTemplateUsageApproval(input.producerDir);
    return;
  }
  const snapshotDigest = input.verdict.metrics?.snapshotDigest;
  if (!input.verdict.ok || snapshotDigest !== input.templateUsage.digest
      || !input.operatorIntentVerdict.ok) {
    throw new Error("template usage gate did not pass against its controller-bound history");
  }
  const foundHistory = historyDigest(input.producerDir, input.templateUsage.path);
  if (foundHistory !== input.templateUsage.digest) {
    throw new Error("template usage gate history no longer matches its authority");
  }
  const core = {
    schemaVersion: 1 as const,
    kind: "producer-template-usage-approval" as const,
    ...content(input),
    historyPath: input.templateUsage.path,
    historyDigest: foundHistory,
    approvedAt: new Date().toISOString(),
  };
  atomicWriteJsonSync(templateUsageApprovalPath(input.producerDir), {
    ...core, digest: hash(core),
  });
}

function validReceipt(value: unknown): value is TemplateUsageApproval {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const row = value as Partial<TemplateUsageApproval>;
  const hashes = [row.planHash, row.planContentHash, row.manifestHash,
    row.transcriptDigest, row.contractHash, row.operatorIntentDigest,
    row.operatorIntentContractHash, row.historyDigest, row.digest];
  return row.schemaVersion === 1 && row.kind === "producer-template-usage-approval"
    && typeof row.historyPath === "string" && path.isAbsolute(row.historyPath)
    && typeof row.approvedAt === "string" && Number.isFinite(Date.parse(row.approvedAt))
    && hashes.every((item) => typeof item === "string" && SHA.test(item));
}

/** Produced/full delivery is impossible without current, tamper-evident history governance. */
export function assertTemplateUsageApprovalCurrent(input: DeliveryInput): void {
  if (!templateUsageApprovalRequired(input.producerDir)) return;
  const destination = templateUsageApprovalPath(input.producerDir);
  let receipt: TemplateUsageApproval;
  try {
    const value = JSON.parse(readFileSync(destination, "utf8")) as unknown;
    if (!validReceipt(value)) throw new Error("invalid receipt shape");
    receipt = value;
  } catch {
    throw new Error("produced/full delivery needs a current template-history review; review the saved plan first");
  }
  const { digest, ...core } = receipt;
  const current = content(input);
  const foundHistory = historyDigest(input.producerDir, receipt.historyPath);
  const currentValues = Object.entries(current) as Array<[keyof typeof current, string]>;
  if (digest !== hash(core) || foundHistory !== receipt.historyDigest
      || currentValues.some(([key, value]) => receipt[key] !== value)) {
    throw new Error("template-history approval is stale; review the current saved plan before delivery");
  }
}
